src/Agent.Plugins/PipelineArtifact/FileShareProvider.cs

```diff
@@ -0,0 +1,97 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Agent.Sdk;
+using Microsoft.TeamFoundation.Build.WebApi;
+using Microsoft.VisualStudio.Services.BlobStore.Common;
+using Microsoft.VisualStudio.Services.Content.Common.Tracing;
+using Microsoft.VisualStudio.Services.Agent.Util;
+
+namespace Agent.Plugins.PipelineArtifact
+{
+    internal class FileShareProvider: IArtifactProvider
+    {
+        private readonly AgentTaskPluginExecutionContext context;
+        private readonly CallbackAppTraceSource tracer;
+
+        public FileShareProvider(AgentTaskPluginExecutionContext context, CallbackAppTraceSource tracer)
+        {
+            this.context = context;
+            this.tracer = tracer;
+        }
+
+        public Task DownloadSingleArtifactAsync(PipelineArtifactDownloadParameters downloadParameters, BuildArtifact buildArtifact, CancellationToken cancellationToken)
+        {
+            return this.DownloadFileShareAsync(downloadParameters.ProjectId, buildArtifact, downloadParameters.TargetDirectory, downloadParameters.MinimatchFilters, cancellationToken);
+        }
+
+        public Task DownloadMultipleArtifactsAsync(PipelineArtifactDownloadParameters downloadParameters, IEnumerable<BuildArtifact> buildArtifacts, CancellationToken cancellationToken)
+        {
+            return this.DownloadFileShareAsync(downloadParameters.ProjectId, buildArtifacts, downloadParameters.TargetDirectory, downloadParameters.MinimatchFilters, cancellationToken);
+        }
+
+        public Task DownloadFileShareAsync(Guid projectId, IEnumerable<BuildArtifact> buildArtifacts, string targetDirectory, IEnumerable<string> minimatchFilters, CancellationToken cancellationToken)
+        {
+            foreach (var buildArtifact in buildArtifacts)
+            {
+                var dirPath = Path.Combine(targetDirectory, buildArtifact.Name);
+                return DownloadFileShareAsync(projectId, buildArtifact, dirPath, minimatchFilters, cancellationToken, isSingleArtifactDownload: false);
```

Comment:
This will terminate on the first file share artifact found. This isn't what we want right? If there are multiple artifacts we don't want to multi download?